File: L3_Arrays&Hashing_Part3/python/SubArraySumEqualsK.py
class Solution(object):
    def subarraySum(self, nums, k):
        count = 0
        sum = 0
        map = defaultdict(int)
        map[0] = 1
        
        for num in nums:
            sum += num
            
            if (sum - k) in map:
                count += map[sum - k]
            
            map[sum] += 1
        
        return count

File: L3_Arrays&Hashing_Part3/cpp/SubArraySumEqualsK.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
      int count = 0, sum = 0;
        std::unordered_map<int, int> map;
        map[0] = 1;
        
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            
            if (map.count(sum - k))
                count += map[sum - k];
            
            map[sum]++;
        }
        
        return count;  
    }
};
